<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>KAIST Design, Build - 2&nbsp; Design</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./index.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./design.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Design</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">KAIST Design, Build</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Home</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./design.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Design</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#environment" id="toc-environment" class="nav-link active" data-scroll-target="#environment"><span class="header-section-number">2.1</span> Environment</a></li>
  <li><a href="#biopython" id="toc-biopython" class="nav-link" data-scroll-target="#biopython"><span class="header-section-number">2.2</span> Biopython</a></li>
  <li><a href="#primers" id="toc-primers" class="nav-link" data-scroll-target="#primers"><span class="header-section-number">2.3</span> Primers</a></li>
  <li><a href="#pydna" id="toc-pydna" class="nav-link" data-scroll-target="#pydna"><span class="header-section-number">2.4</span> pydna</a></li>
  <li><a href="#parts" id="toc-parts" class="nav-link" data-scroll-target="#parts"><span class="header-section-number">2.5</span> Parts</a></li>
  <li><a href="#feature-amplification" id="toc-feature-amplification" class="nav-link" data-scroll-target="#feature-amplification"><span class="header-section-number">2.6</span> Feature amplification</a>
  <ul class="collapse">
  <li><a href="#primer-design" id="toc-primer-design" class="nav-link" data-scroll-target="#primer-design"><span class="header-section-number">2.6.1</span> primer design</a></li>
  <li><a href="#pcr" id="toc-pcr" class="nav-link" data-scroll-target="#pcr"><span class="header-section-number">2.6.2</span> PCR</a></li>
  </ul></li>
  <li><a href="#golden-gate-assembly-i" id="toc-golden-gate-assembly-i" class="nav-link" data-scroll-target="#golden-gate-assembly-i"><span class="header-section-number">2.7</span> Golden Gate Assembly I</a>
  <ul class="collapse">
  <li><a href="#enzyme-cutting" id="toc-enzyme-cutting" class="nav-link" data-scroll-target="#enzyme-cutting"><span class="header-section-number">2.7.1</span> Enzyme cutting</a></li>
  <li><a href="#assembly-the-fragments-back" id="toc-assembly-the-fragments-back" class="nav-link" data-scroll-target="#assembly-the-fragments-back"><span class="header-section-number">2.7.2</span> Assembly the fragments back</a></li>
  </ul></li>
  <li><a href="#golden-gate-assembly-ii" id="toc-golden-gate-assembly-ii" class="nav-link" data-scroll-target="#golden-gate-assembly-ii"><span class="header-section-number">2.8</span> Golden Gate Assembly II</a>
  <ul class="collapse">
  <li><a href="#find-positions-of-a-specific-feature-label" id="toc-find-positions-of-a-specific-feature-label" class="nav-link" data-scroll-target="#find-positions-of-a-specific-feature-label"><span class="header-section-number">2.8.1</span> Find positions of a specific feature label</a></li>
  <li><a href="#part-preparation-by-pcr" id="toc-part-preparation-by-pcr" class="nav-link" data-scroll-target="#part-preparation-by-pcr"><span class="header-section-number">2.8.2</span> Part preparation by PCR</a></li>
  <li><a href="#enzyme-cut-and-assembly" id="toc-enzyme-cut-and-assembly" class="nav-link" data-scroll-target="#enzyme-cut-and-assembly"><span class="header-section-number">2.8.3</span> Enzyme cut and assembly</a></li>
  </ul></li>
  <li><a href="#assembly-with-wrong-fragments" id="toc-assembly-with-wrong-fragments" class="nav-link" data-scroll-target="#assembly-with-wrong-fragments"><span class="header-section-number">2.9</span> Assembly with wrong fragments</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Design</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="environment" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="environment"><span class="header-section-number">2.1</span> Environment</h2>
<ul>
<li><p>conda</p></li>
<li><p>docker</p></li>
<li><p>colab</p></li>
<li><p>To make google drive as colab working directory</p></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> google.colab <span class="im">import</span> drive</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>drive.mount(<span class="st">'/content/drive'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>Change working directory</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>os.chdir(<span class="st">'/content/drive/MyDrive/design_build'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="shell"><code>!pwd
!wget https://media.addgene.org/snapgene-media/v1.7.9-0-g88a3305/sequences/222046/51c2cfab-a3b4-4d62-98df-0c77ec21164e/addgene-plasmid-50005-sequence-222046.gbk</code></pre>
<div id="cell-5" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ls</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>README.md    data      docs    index.qmd
_quarto.yml  design.ipynb  images  seq_analysis.ipynb</code></pre>
</div>
</div>
<p><img src="images/puc19.png" class="img-fluid" width="500"></p>
</section>
<section id="biopython" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="biopython"><span class="header-section-number">2.2</span> Biopython</h2>
<p><a href="https://biopython.org/">Biopython</a> is a collection of freely available Python tools for computational biology and bioinformatics.</p>
<div id="cell-9" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> Bio <span class="im">import</span> SeqIO</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pandas <span class="im">import</span> DataFrame</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>records <span class="op">=</span> SeqIO.read(<span class="st">"data/parts/addgene-plasmid-50005-sequence-222046.gbk"</span>, <span class="st">"genbank"</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>features <span class="op">=</span> []</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> feature <span class="kw">in</span> records.features:</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    features.append({</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Label"</span>: feature.qualifiers.get(<span class="st">"label"</span>, [<span class="st">""</span>])[<span class="dv">0</span>],</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Strand"</span>: feature.strand,</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Start"</span>: feature.location.start,</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"End"</span>: feature.location.end, </span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Type"</span>: feature.<span class="bu">type</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(features)</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>DataFrame(features)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[{'Label': '', 'Strand': 1, 'Start': ExactPosition(0), 'End': ExactPosition(2686), 'Type': 'source'}, {'Label': 'pBR322ori-F', 'Strand': 1, 'Start': ExactPosition(117), 'End': ExactPosition(137), 'Type': 'primer_bind'}, {'Label': 'L4440', 'Strand': 1, 'Start': ExactPosition(370), 'End': ExactPosition(388), 'Type': 'primer_bind'}, {'Label': 'CAP binding site', 'Strand': 1, 'Start': ExactPosition(504), 'End': ExactPosition(526), 'Type': 'protein_bind'}, {'Label': 'lac promoter', 'Strand': 1, 'Start': ExactPosition(540), 'End': ExactPosition(571), 'Type': 'promoter'}, {'Label': 'lac operator', 'Strand': 1, 'Start': ExactPosition(578), 'End': ExactPosition(595), 'Type': 'protein_bind'}, {'Label': 'M13/pUC Reverse', 'Strand': 1, 'Start': ExactPosition(583), 'End': ExactPosition(606), 'Type': 'primer_bind'}, {'Label': 'M13 rev', 'Strand': 1, 'Start': ExactPosition(602), 'End': ExactPosition(619), 'Type': 'primer_bind'}, {'Label': 'M13 Reverse', 'Strand': 1, 'Start': ExactPosition(602), 'End': ExactPosition(619), 'Type': 'primer_bind'}, {'Label': 'lacZ-alpha', 'Strand': 1, 'Start': ExactPosition(614), 'End': ExactPosition(938), 'Type': 'CDS'}, {'Label': 'MCS', 'Strand': 1, 'Start': ExactPosition(631), 'End': ExactPosition(688), 'Type': 'misc_feature'}, {'Label': 'M13 Forward', 'Strand': -1, 'Start': ExactPosition(688), 'End': ExactPosition(706), 'Type': 'primer_bind'}, {'Label': 'M13 fwd', 'Strand': -1, 'Start': ExactPosition(688), 'End': ExactPosition(705), 'Type': 'primer_bind'}, {'Label': 'M13/pUC Forward', 'Strand': -1, 'Start': ExactPosition(697), 'End': ExactPosition(720), 'Type': 'primer_bind'}, {'Label': 'pRS-marker', 'Strand': -1, 'Start': ExactPosition(913), 'End': ExactPosition(933), 'Type': 'primer_bind'}, {'Label': "pGEX 3'", 'Strand': 1, 'Start': ExactPosition(1032), 'End': ExactPosition(1055), 'Type': 'primer_bind'}, {'Label': 'pBRforEco', 'Strand': -1, 'Start': ExactPosition(1092), 'End': ExactPosition(1111), 'Type': 'primer_bind'}, {'Label': 'AmpR promoter', 'Strand': 1, 'Start': ExactPosition(1178), 'End': ExactPosition(1283), 'Type': 'promoter'}, {'Label': 'AmpR', 'Strand': 1, 'Start': ExactPosition(1283), 'End': ExactPosition(2144), 'Type': 'CDS'}, {'Label': 'Amp-R', 'Strand': -1, 'Start': ExactPosition(1501), 'End': ExactPosition(1521), 'Type': 'primer_bind'}, {'Label': 'ori', 'Strand': 1, 'Start': ExactPosition(0), 'End': ExactPosition(2686), 'Type': 'rep_origin'}]</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/haseong/anaconda3/envs/biopy/lib/python3.11/site-packages/Bio/SeqFeature.py:1040: BiopythonParserWarning: Attempting to fix invalid location '2315..217' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.
  warnings.warn(
/home/haseong/anaconda3/envs/biopy/lib/python3.11/site-packages/Bio/SeqFeature.py:230: BiopythonDeprecationWarning: Please use .location.strand rather than .strand
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="4">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Label</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td></td>
<td>1</td>
<td>0</td>
<td>2686</td>
<td>source</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>pBR322ori-F</td>
<td>1</td>
<td>117</td>
<td>137</td>
<td>primer_bind</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>L4440</td>
<td>1</td>
<td>370</td>
<td>388</td>
<td>primer_bind</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>CAP binding site</td>
<td>1</td>
<td>504</td>
<td>526</td>
<td>protein_bind</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>lac promoter</td>
<td>1</td>
<td>540</td>
<td>571</td>
<td>promoter</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>lac operator</td>
<td>1</td>
<td>578</td>
<td>595</td>
<td>protein_bind</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>M13/pUC Reverse</td>
<td>1</td>
<td>583</td>
<td>606</td>
<td>primer_bind</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>M13 rev</td>
<td>1</td>
<td>602</td>
<td>619</td>
<td>primer_bind</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>M13 Reverse</td>
<td>1</td>
<td>602</td>
<td>619</td>
<td>primer_bind</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>lacZ-alpha</td>
<td>1</td>
<td>614</td>
<td>938</td>
<td>CDS</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>MCS</td>
<td>1</td>
<td>631</td>
<td>688</td>
<td>misc_feature</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">11</td>
<td>M13 Forward</td>
<td>-1</td>
<td>688</td>
<td>706</td>
<td>primer_bind</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">12</td>
<td>M13 fwd</td>
<td>-1</td>
<td>688</td>
<td>705</td>
<td>primer_bind</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">13</td>
<td>M13/pUC Forward</td>
<td>-1</td>
<td>697</td>
<td>720</td>
<td>primer_bind</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">14</td>
<td>pRS-marker</td>
<td>-1</td>
<td>913</td>
<td>933</td>
<td>primer_bind</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">15</td>
<td>pGEX 3'</td>
<td>1</td>
<td>1032</td>
<td>1055</td>
<td>primer_bind</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">16</td>
<td>pBRforEco</td>
<td>-1</td>
<td>1092</td>
<td>1111</td>
<td>primer_bind</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">17</td>
<td>AmpR promoter</td>
<td>1</td>
<td>1178</td>
<td>1283</td>
<td>promoter</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">18</td>
<td>AmpR</td>
<td>1</td>
<td>1283</td>
<td>2144</td>
<td>CDS</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">19</td>
<td>Amp-R</td>
<td>-1</td>
<td>1501</td>
<td>1521</td>
<td>primer_bind</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">20</td>
<td>ori</td>
<td>1</td>
<td>0</td>
<td>2686</td>
<td>rep_origin</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</div>
</section>
<section id="primers" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="primers"><span class="header-section-number">2.3</span> Primers</h2>
<p><a href="https://github.com/Lattice-Automation/primers">primers</a> It is uniquely focused on DNA assembly flows like Gibson Assembly and Golden Gate cloning. You can design primers while adding sequence to the 5’ ends of primers.</p>
<div id="cell-12" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> primers <span class="im">import</span> create, primers</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pandas <span class="im">import</span> DataFrame</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> random <span class="im">import</span> sample, choices</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>myseq_list <span class="op">=</span> choices([<span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"G"</span>, <span class="st">"C"</span>], k<span class="op">=</span><span class="dv">100</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>myseq <span class="op">=</span> <span class="st">""</span>.join(myseq_list)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(myseq)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>fwd, rev <span class="op">=</span> create(myseq, add_fwd <span class="op">=</span> <span class="st">"GGGG"</span>, add_rev <span class="op">=</span> <span class="st">"TTTT"</span>)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co"># p1, p2 = primers(myseq, add_fwd = "GGGG", add_rev = "TTTT")</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fwd)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(rev)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="co">## display table form</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>DataFrame(<span class="bu">list</span>(fwd.<span class="bu">dict</span>().values())[:<span class="op">-</span><span class="dv">1</span>], index <span class="op">=</span> <span class="bu">list</span>(rev.<span class="bu">dict</span>().keys())[:<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a><span class="co">## default argument values </span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>create</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CGTGCCTCCATAAATAACTTGCAAGATTCTCACCATTCGAAGGTTCTCGACAAGGGGCGGGGGGTAAAAATAGCATTACTAGTTCGGATAAATCTGCCCT
Primer(seq='GGGGCGTGCCTCCATAAATAACTTG', len=25, tm=63.3, tm_total=70.6, gc=0.5, dg=0, fwd=True, off_target_count=0, scoring=Scoring(penalty=1.8, penalty_tm=1.3, penalty_tm_diff=0, penalty_gc=0.0, penalty_len=0.5, penalty_dg=0.0, penalty_off_target=0.0))
Primer(seq='TTTTAGGGCAGATTTATCCGAACTAGT', len=27, tm=63.4, tm_total=63.9, gc=0.4, dg=-0.37, fwd=False, off_target_count=0, scoring=Scoring(penalty=4.6, penalty_tm=1.4, penalty_tm_diff=0, penalty_gc=2.0, penalty_len=0.5, penalty_dg=0.7, penalty_off_target=0.0))</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>&lt;function primers.primers.primers(seq: str, add_fwd: str = '', add_rev: str = '', add_fwd_len: Tuple[int, int] = (-1, -1), add_rev_len: Tuple[int, int] = (-1, -1), offtarget_check: str = '', optimal_tm: float = 62.0, optimal_gc: float = 0.5, optimal_len: int = 22, penalty_tm: float = 1.0, penalty_gc: float = 0.2, penalty_len: float = 0.5, penalty_tm_diff: float = 1.0, penalty_dg: float = 2.0, penalty_off_target: float = 20.0) -&gt; Tuple[primers.primers.Primer, primers.primers.Primer]&gt;</code></pre>
</div>
</div>
<ul>
<li>Default arguments and values</li>
</ul>
<pre class="plain"><code>&lt;function primers.primers.primers(seq: str, add_fwd: str = '', add_rev: str = '', add_fwd_len: Tuple[int, int] = (-1, -1), add_rev_len: Tuple[int, int] = (-1, -1), offtarget_check: str = '', optimal_tm: float = 62.0, optimal_gc: float = 0.5, optimal_len: int = 22, penalty_tm: float = 1.0, penalty_gc: float = 0.2, penalty_len: float = 0.5, penalty_tm_diff: float = 1.0, penalty_dg: float = 2.0, penalty_off_target: float = 20.0) -&gt; Tuple[primers.primers.Primer, primers.primers.Primer]&gt;</code></pre>
<ul>
<li>offtarget</li>
</ul>
<div id="cell-15" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> primers <span class="im">import</span> create</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> random <span class="im">import</span> choices</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> print_primer_info(x):</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    <span class="im">from</span> pandas <span class="im">import</span> DataFrame</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> DataFrame(<span class="bu">list</span>(x.<span class="bu">dict</span>().values())[:<span class="op">-</span><span class="dv">1</span>], index <span class="op">=</span> <span class="bu">list</span>(x.<span class="bu">dict</span>().keys())[:<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(df)</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>primer_binding_seq <span class="op">=</span> <span class="st">"GTCATATGCATTCGATGCGTTAGG"</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>rnd_seq1 <span class="op">=</span> <span class="st">""</span>.join(choices([<span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"G"</span>, <span class="st">"C"</span>], k<span class="op">=</span><span class="dv">100</span>))</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>rnd_seq2 <span class="op">=</span> <span class="st">""</span>.join(choices([<span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"G"</span>, <span class="st">"C"</span>], k<span class="op">=</span><span class="dv">100</span>))</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>myseq <span class="op">=</span> primer_binding_seq<span class="op">+</span>rnd_seq1</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(myseq)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(myseq)</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>fwd, rev <span class="op">=</span> create(myseq)</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>print_primer_info(fwd)</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a><span class="co">## primer considering offtargets</span></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>myseq2 <span class="op">=</span> primer_binding_seq<span class="op">+</span>rnd_seq1<span class="op">+</span>primer_binding_seq<span class="op">+</span>rnd_seq2</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>fwd2, rev <span class="op">=</span> create(myseq2)</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>print_primer_info(fwd2)</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a><span class="co">## optimal_tm is ignored </span></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>fwd2, rev <span class="op">=</span> create(myseq2, optimal_tm <span class="op">=</span> <span class="dv">62</span>)</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>print_primer_info(fwd2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GTCATATGCATTCGATGCGTTAGGGACACCCATGGCAACATGTGGATATAACTCGGTGCTGAGGAAAACTTCATACGCTCTTGACGTTCTATGCATGAAGCCCTTTCAACGCACGCTTCATTCA
                                     0
seq               GTCATATGCATTCGATGCGT
len                                 20
tm                                62.5
tm_total                          62.5
gc                                 0.5
dg                               -0.56
fwd                               True
off_target_count                     0
                                           0
seq               GTCATATGCATTCGATGCGTTAGGGA
len                                       26
tm                                      68.0
tm_total                                68.0
gc                                       0.5
dg                                     -0.56
fwd                                     True
off_target_count                           0
                                           0
seq               GTCATATGCATTCGATGCGTTAGGGA
len                                       26
tm                                      68.0
tm_total                                68.0
gc                                       0.5
dg                                     -0.56
fwd                                     True
off_target_count                           0</code></pre>
</div>
</div>
</section>
<section id="pydna" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="pydna"><span class="header-section-number">2.4</span> pydna</h2>
<ul>
<li><a href="https://github.com/BjornFJohansson/pydna">pyDNA</a> The pydna python package provide a human-readable formal descriptions of 🧬 cloning and genetic assembly strategies in Python 🐍 which allow for simulation and verification.</li>
</ul>
<div id="cell-17" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.dseqrecord <span class="im">import</span> Dseqrecord</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>dsr <span class="op">=</span> Dseqrecord(<span class="st">"ATGCGTTGC"</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>dsr.figure()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<div class="ansi-escaped-output">
<pre>Dseqrecord(-9)
ATGCGTTGC
TACGCAACG</pre>
</div>
</div>
</div>
<div id="cell-18" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.readers <span class="im">import</span> read</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> read(<span class="st">"data/parts/addgene-plasmid-50005-sequence-222046.gbk"</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>p.list_features()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/haseong/anaconda3/envs/biopy/lib/python3.11/site-packages/Bio/SeqFeature.py:1040: BiopythonParserWarning: Attempting to fix invalid location '2315..217' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="9">
<div>
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Ft#</th>
<th data-quarto-table-cell-role="th">Label or Note</th>
<th data-quarto-table-cell-role="th">Dir</th>
<th data-quarto-table-cell-role="th">Sta</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Len</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">orf?</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>nd</td>
<td>--&gt;</td>
<td>0</td>
<td>2686</td>
<td>2686</td>
<td>source</td>
<td>no</td>
</tr>
<tr class="even">
<td>1</td>
<td>L:pBR322ori-F</td>
<td>--&gt;</td>
<td>117</td>
<td>137</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>2</td>
<td>L:L4440</td>
<td>--&gt;</td>
<td>370</td>
<td>388</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>3</td>
<td>L:CAP binding si</td>
<td>--&gt;</td>
<td>504</td>
<td>526</td>
<td>22</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>4</td>
<td>L:lac promoter</td>
<td>--&gt;</td>
<td>540</td>
<td>571</td>
<td>31</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="even">
<td>5</td>
<td>L:lac operator</td>
<td>--&gt;</td>
<td>578</td>
<td>595</td>
<td>17</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>6</td>
<td>L:M13/pUC Revers</td>
<td>--&gt;</td>
<td>583</td>
<td>606</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>7</td>
<td>L:M13 rev</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>8</td>
<td>L:M13 Reverse</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>9</td>
<td>L:lacZ-alpha</td>
<td>--&gt;</td>
<td>614</td>
<td>938</td>
<td>324</td>
<td>CDS</td>
<td>yes</td>
</tr>
<tr class="odd">
<td>10</td>
<td>L:MCS</td>
<td>--&gt;</td>
<td>631</td>
<td>688</td>
<td>57</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>11</td>
<td>L:M13 Forward</td>
<td>&lt;--</td>
<td>688</td>
<td>706</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>12</td>
<td>L:M13 fwd</td>
<td>&lt;--</td>
<td>688</td>
<td>705</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>13</td>
<td>L:M13/pUC Forwar</td>
<td>&lt;--</td>
<td>697</td>
<td>720</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>14</td>
<td>L:pRS-marker</td>
<td>&lt;--</td>
<td>913</td>
<td>933</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>15</td>
<td>L:pGEX 3'</td>
<td>--&gt;</td>
<td>1032</td>
<td>1055</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>16</td>
<td>L:pBRforEco</td>
<td>&lt;--</td>
<td>1092</td>
<td>1111</td>
<td>19</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>17</td>
<td>L:AmpR promoter</td>
<td>--&gt;</td>
<td>1178</td>
<td>1283</td>
<td>105</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="odd">
<td>18</td>
<td>L:AmpR</td>
<td>--&gt;</td>
<td>1283</td>
<td>2144</td>
<td>861</td>
<td>CDS</td>
<td>yes</td>
</tr>
<tr class="even">
<td>19</td>
<td>L:Amp-R</td>
<td>&lt;--</td>
<td>1501</td>
<td>1521</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>20</td>
<td>L:ori</td>
<td>--&gt;</td>
<td>0</td>
<td>2686</td>
<td>589</td>
<td>rep_origin</td>
<td>no</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="cell-19" class="cell" data-execution_count="123">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>extracted_site <span class="op">=</span> p.extract_feature(<span class="dv">10</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>extracted_site.seq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="123">
<pre><code>Dseq(-57)
AAGC..ATTC
TTCG..TAAG</code></pre>
</div>
</div>
</section>
<section id="parts" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="parts"><span class="header-section-number">2.5</span> Parts</h2>
<ul>
<li><p>pUC19 from Addgene</p></li>
<li><p>Remove BsaI site</p></li>
<li><p>Insert a part into MCS</p></li>
<li><p>Hinz, Aaron J., Benjamin Stenzler, and Alexandre J. Poulain. “Golden gate assembly of aerobic and anaerobic microbial bioreporters.” Applied and environmental microbiology 88.1 (2022): e01485-21.</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/pUC19.png" class="img-fluid figure-img" width="500"></p>
<figcaption>plasmid pUC19</figcaption>
</figure>
</div>
<p><img src="images/amp_bsaI1.png" height="200" alt="bsaI replacement"> <img src="images/amp_bsaI2.png" height="200" alt="bsaI replacement2"></p>
<section id="list-of-parts" class="level4" data-number="2.5.0.1">
<h4 data-number="2.5.0.1" class="anchored" data-anchor-id="list-of-parts"><span class="header-section-number">2.5.0.1</span> List of parts</h4>
<ul>
<li>pUC19-J23100.gb</li>
<li>pUC19-RB0030.gb</li>
<li>pUC19-L2U3H03.gb</li>
<li>pUC19-egfp.gb</li>
</ul>
<p><img src="images/puc19_egfp.png" class="img-fluid" width="500"></p>
</section>
</section>
<section id="feature-amplification" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="feature-amplification"><span class="header-section-number">2.6</span> Feature amplification</h2>
<section id="primer-design" class="level3" data-number="2.6.1">
<h3 data-number="2.6.1" class="anchored" data-anchor-id="primer-design"><span class="header-section-number">2.6.1</span> primer design</h3>
<div id="cell-27" class="cell" data-execution_count="148">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.readers <span class="im">import</span> read</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> primers <span class="im">import</span> create</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>egfp <span class="op">=</span> read(<span class="st">"data/parts/pUC19-egfp.gb"</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>egfp_feature_list <span class="op">=</span> egfp.list_features()</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>display(egfp_feature_list)</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>myseq <span class="op">=</span> egfp.extract_feature(<span class="dv">13</span>)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="bu">str</span>(myseq.seq)</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>fwd, rev <span class="op">=</span> create(<span class="bu">str</span>(myseq.seq), add_fwd <span class="op">=</span> <span class="st">"GGGG"</span>, add_rev <span class="op">=</span> <span class="st">"TTTT"</span>)</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(fwd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/haseong/anaconda3/envs/biopy/lib/python3.11/site-packages/Bio/GenBank/Scanner.py:1528: BiopythonParserWarning: Attempting to parse malformed locus line:
'LOCUS       pUC19-egfp        3371 bp DNA     circular SYN 18-MAY-2024\n'
Found locus 'pUC19-egfp' size '3371' residue_type 'DNA'
Some fields may be wrong.
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Ft#</th>
<th data-quarto-table-cell-role="th">Label or Note</th>
<th data-quarto-table-cell-role="th">Dir</th>
<th data-quarto-table-cell-role="th">Sta</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Len</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">orf?</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>nd</td>
<td>--&gt;</td>
<td>0</td>
<td>3371</td>
<td>3371</td>
<td>source</td>
<td>no</td>
</tr>
<tr class="even">
<td>1</td>
<td>L:pBR322ori-F</td>
<td>--&gt;</td>
<td>117</td>
<td>137</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>2</td>
<td>L:L4440</td>
<td>--&gt;</td>
<td>370</td>
<td>388</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>3</td>
<td>L:CAP binding si</td>
<td>--&gt;</td>
<td>504</td>
<td>526</td>
<td>22</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>4</td>
<td>L:lac promoter</td>
<td>--&gt;</td>
<td>540</td>
<td>571</td>
<td>31</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="even">
<td>5</td>
<td>L:lac operator</td>
<td>--&gt;</td>
<td>578</td>
<td>595</td>
<td>17</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>6</td>
<td>L:M13/pUC Revers</td>
<td>--&gt;</td>
<td>583</td>
<td>606</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>7</td>
<td>L:M13 rev</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>8</td>
<td>L:M13 Reverse</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>9</td>
<td>L:lacZ-alpha</td>
<td>--&gt;</td>
<td>614</td>
<td>1623</td>
<td>1009</td>
<td>CDS</td>
<td>no</td>
</tr>
<tr class="odd">
<td>10</td>
<td>L:BsaI-F</td>
<td>--&gt;</td>
<td>631</td>
<td>638</td>
<td>7</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>11</td>
<td>L:lacZ-alpha</td>
<td>--&gt;</td>
<td>638</td>
<td>642</td>
<td>4</td>
<td>CDS</td>
<td>no</td>
</tr>
<tr class="odd">
<td>12</td>
<td>L:oh-f2-f3-F</td>
<td>--&gt;</td>
<td>638</td>
<td>642</td>
<td>4</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>13</td>
<td>L:EGFP</td>
<td>--&gt;</td>
<td>642</td>
<td>1362</td>
<td>720</td>
<td>CDS</td>
<td>yes</td>
</tr>
<tr class="odd">
<td>14</td>
<td>L:EGFP</td>
<td>--&gt;</td>
<td>642</td>
<td>1359</td>
<td>717</td>
<td>CDS</td>
<td>no</td>
</tr>
<tr class="even">
<td>15</td>
<td>L:EGFP-N</td>
<td>&lt;--</td>
<td>687</td>
<td>709</td>
<td>22</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>16</td>
<td>L:EXFP-R</td>
<td>&lt;--</td>
<td>948</td>
<td>968</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>17</td>
<td>L:EGFP-C</td>
<td>--&gt;</td>
<td>1295</td>
<td>1317</td>
<td>22</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>18</td>
<td>L:oh-f3-f4-F</td>
<td>--&gt;</td>
<td>1362</td>
<td>1366</td>
<td>4</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>19</td>
<td>L:BasI-R</td>
<td>--&gt;</td>
<td>1366</td>
<td>1373</td>
<td>7</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="odd">
<td>20</td>
<td>L:M13 Forward</td>
<td>&lt;--</td>
<td>1373</td>
<td>1391</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>21</td>
<td>L:M13 fwd</td>
<td>&lt;--</td>
<td>1373</td>
<td>1390</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>22</td>
<td>L:M13/pUC Forwar</td>
<td>&lt;--</td>
<td>1382</td>
<td>1405</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>23</td>
<td>L:pRS-marker</td>
<td>&lt;--</td>
<td>1598</td>
<td>1618</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>24</td>
<td>L:pGEX 3'</td>
<td>--&gt;</td>
<td>1717</td>
<td>1740</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>25</td>
<td>L:pBRforEco</td>
<td>&lt;--</td>
<td>1777</td>
<td>1796</td>
<td>19</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>26</td>
<td>L:AmpR promoter</td>
<td>--&gt;</td>
<td>1863</td>
<td>1968</td>
<td>105</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="even">
<td>27</td>
<td>L:AmpR</td>
<td>--&gt;</td>
<td>1968</td>
<td>2829</td>
<td>861</td>
<td>CDS</td>
<td>yes</td>
</tr>
<tr class="odd">
<td>28</td>
<td>L:Amp-R</td>
<td>&lt;--</td>
<td>2186</td>
<td>2206</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>29</td>
<td>L:ori</td>
<td>--&gt;</td>
<td>0</td>
<td>3371</td>
<td>589</td>
<td>rep_origin</td>
<td>no</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Primer(seq='GGGGATGGTGAGCAAGGGCG', len=20, tm=65.1, tm_total=72.1, gc=0.7, dg=0, fwd=True, off_target_count=0, scoring=Scoring(penalty=10.1, penalty_tm=3.1, penalty_tm_diff=0, penalty_gc=4.0, penalty_len=3.0, penalty_dg=0.0, penalty_off_target=0.0))</code></pre>
</div>
</div>
</section>
<section id="pcr" class="level3" data-number="2.6.2">
<h3 data-number="2.6.2" class="anchored" data-anchor-id="pcr"><span class="header-section-number">2.6.2</span> PCR</h3>
<div id="cell-29" class="cell" data-execution_count="149">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.<span class="bu">all</span> <span class="im">import</span> pcr</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>pcr_product <span class="op">=</span> pcr(fwd.seq, rev.seq, egfp.extract_feature(<span class="dv">13</span>))</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>pcr_product.figure()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="149">
<pre><code>    5ATGGTGAGCAAGGGCG...CATGGACGAGCTGTACAAGTAA3
                        ||||||||||||||||||||||
                       3GTACCTGCTCGACATGTTCATTTTTT5
5GGGGATGGTGAGCAAGGGCG3
     ||||||||||||||||
    3TACCACTCGTTCCCGC...GTACCTGCTCGACATGTTCATT5</code></pre>
</div>
</div>
</section>
</section>
<section id="golden-gate-assembly-i" class="level2" data-number="2.7">
<h2 data-number="2.7" class="anchored" data-anchor-id="golden-gate-assembly-i"><span class="header-section-number">2.7</span> Golden Gate Assembly I</h2>
<ul>
<li>Target sequence including overhang and BsaI site</li>
<li>Generate primers</li>
<li>PCR the target sequence</li>
</ul>
<div id="cell-31" class="cell" data-execution_count="152">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> primers <span class="im">import</span> create</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.<span class="bu">all</span> <span class="im">import</span> pcr</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>frag1 <span class="op">=</span> egfp.extract_feature(<span class="dv">10</span>)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>frag2 <span class="op">=</span> egfp.extract_feature(<span class="dv">12</span>)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>frag3 <span class="op">=</span> egfp.extract_feature(<span class="dv">13</span>)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>frag4 <span class="op">=</span> egfp.extract_feature(<span class="dv">18</span>)</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>frag5 <span class="op">=</span> egfp.extract_feature(<span class="dv">19</span>)</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>targetseq <span class="op">=</span> frag1<span class="op">+</span>frag2<span class="op">+</span>frag3<span class="op">+</span>frag4<span class="op">+</span>frag5</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>fwd, rev <span class="op">=</span> create(<span class="bu">str</span>(targetseq.seq))</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>pcr_product <span class="op">=</span> pcr(fwd.seq, rev.seq, targetseq)</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>pcr_product.figure()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="152">
<pre><code>5GGTCTCAGTCAATGGTGA...TACAAGTAAGGGATGAGACC3
                      ||||||||||||||||||||
                     3ATGTTCATTCCCTACTCTGG5
5GGTCTCAGTCAATGGTGA3
 ||||||||||||||||||
3CCAGAGTCAGTTACCACT...ATGTTCATTCCCTACTCTGG5</code></pre>
</div>
</div>
<section id="enzyme-cutting" class="level3" data-number="2.7.1">
<h3 data-number="2.7.1" class="anchored" data-anchor-id="enzyme-cutting"><span class="header-section-number">2.7.1</span> Enzyme cutting</h3>
<div id="cell-33" class="cell" data-execution_count="153">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> Bio.Restriction <span class="im">import</span> BsaI</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>cut_product <span class="op">=</span> pcr_product.cut(BsaI)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(cut_product))</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>display(cut_product[<span class="dv">0</span>].figure())</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>()</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>display(cut_product[<span class="dv">1</span>].figure())</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>()</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>display(cut_product[<span class="dv">2</span>].figure())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>3

</code></pre>
</div>
<div class="cell-output cell-output-display">
<div class="ansi-escaped-output">
<pre>Dseqrecord(-11)
<span class="ansi-bright-yellow-bg">GGTCTCA</span>    
CCAGAGTCAGT</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div class="ansi-escaped-output">
<pre>Dseqrecord(-728)
<span class="ansi-bright-yellow-bg">GTCA</span>ATGGTGAGCAAGGGCGAGGAGCTGTTCACCGGGGTGGTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCGTGCCCTGGCCCACCCTCGTGACCACCCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGCCCGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGCAACTACAAGACCCGCGCCGAGGTGAAGTTCGAGGGCGACACCCTGGTGAACCGCATCGAGCTGAAGGGCATCGACTTCAAGGAGGACGGCAACATCCTGGGGCACAAGCTGGAGTACAACTACAACAGCCACAACGTCTATATCATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCGACGGCCCCGTGCTGCTGCCCGACAACCACTACCTGAGCACCCAGTCCGCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACAAGTAA    
    TACCACTCGTTCCCGCTCCTCGACAAGTGGCCCCACCACGGGTAGGACCAGCTCGACCTGCCGCTGCATTTGCCGGTGTTCAAGTCGCACAGGCCGCTCCCGCTCCCGCTACGGTGGATGCCGTTCGACTGGGACTTCAAGTAGACGTGGTGGCCGTTCGACGGGCACGGGACCGGGTGGGAGCACTGGTGGGACTGGATGCCGCACGTCACGAAGTCGGCGATGGGGCTGGTGTACTTCGTCGTGCTGAAGAAGTTCAGGCGGTACGGGCTTCCGATGCAGGTCCTCGCGTGGTAGAAGAAGTTCCTGCTGCCGTTGATGTTCTGGGCGCGGCTCCACTTCAAGCTCCCGCTGTGGGACCACTTGGCGTAGCTCGACTTCCCGTAGCTGAAGTTCCTCCTGCCGTTGTAGGACCCCGTGTTCGACCTCATGTTGATGTTGTCGGTGTTGCAGATATAGTACCGGCTGTTCGTCTTCTTGCCGTAGTTCCACTTGAAGTTCTAGGCGGTGTTGTAGCTCCTGCCGTCGCACGTCGAGCGGCTGGTGATGGTCGTCTTGTGGGGGTAGCCGCTGCCGGGGCACGACGACGGGCTGTTGGTGATGGACTCGTGGGTCAGGCGGGACTCGTTTCTGGGGTTGCTCTTCGCGCTAGTGTACCAGGACGACCTCAAGCACTGGCGGCGGCCCTAGTGAGAGCCGTACCTGCTCGACATGTTCATTCCCT</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div class="ansi-escaped-output">
<pre>Dseqrecord(-11)
<span class="ansi-bright-yellow-bg">GGGA</span>TGAGACC
    ACTCTGG</pre>
</div>
</div>
</div>
</section>
<section id="assembly-the-fragments-back" class="level3" data-number="2.7.2">
<h3 data-number="2.7.2" class="anchored" data-anchor-id="assembly-the-fragments-back"><span class="header-section-number">2.7.2</span> Assembly the fragments back</h3>
<div id="cell-35" class="cell" data-execution_count="140">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.assembly <span class="im">import</span> Assembly</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.common_sub_strings <span class="im">import</span> terminal_overlap</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>asm <span class="op">=</span> Assembly((cut_product[<span class="dv">0</span>], cut_product[<span class="dv">1</span>], cut_product[<span class="dv">2</span>]), algorithm <span class="op">=</span> terminal_overlap, limit <span class="op">=</span> <span class="dv">4</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>candidate <span class="op">=</span> asm.assemble_linear()</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(candidate[<span class="dv">0</span>])</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>candidate[<span class="dv">0</span>].figure()   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Dseqrecord
circular: False
size: 742
ID: id
Name: name
Description: description
Number of features: 13
/molecule_type=DNA
Dseq(-742)
GGTC..GACC
CCAG..CTGG</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="140">
<pre><code>name| 4
     \/
     /\
      4|name| 4
             \/
             /\
              4|name</code></pre>
</div>
</div>
</section>
</section>
<section id="golden-gate-assembly-ii" class="level2" data-number="2.8">
<h2 data-number="2.8" class="anchored" data-anchor-id="golden-gate-assembly-ii"><span class="header-section-number">2.8</span> Golden Gate Assembly II</h2>
<section id="find-positions-of-a-specific-feature-label" class="level3" data-number="2.8.1">
<h3 data-number="2.8.1" class="anchored" data-anchor-id="find-positions-of-a-specific-feature-label"><span class="header-section-number">2.8.1</span> Find positions of a specific feature label</h3>
<div id="cell-37" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.readers <span class="im">import</span> read</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="co">## read the parts</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>egfp <span class="op">=</span> read(<span class="st">"data/parts/pUC19-egfp.gb"</span>)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>promoter <span class="op">=</span> read(<span class="st">"data/parts/pUC19-J23100.gb"</span>)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>terminator <span class="op">=</span> read(<span class="st">"data/parts/pUC19-L2U3H03.gb"</span>)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>rbs <span class="op">=</span> read(<span class="st">"data/parts/pUC19-RB0030.gb"</span>)</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>display(promoter.list_features())</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>display(rbs.list_features())</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>display(egfp.list_features())</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>display(terminator.list_features())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/haseong/anaconda3/envs/biopy/lib/python3.11/site-packages/Bio/GenBank/Scanner.py:1528: BiopythonParserWarning: Attempting to parse malformed locus line:
'LOCUS       pUC19-egfp        3371 bp DNA     circular SYN 18-MAY-2024\n'
Found locus 'pUC19-egfp' size '3371' residue_type 'DNA'
Some fields may be wrong.
  warnings.warn(
/home/haseong/anaconda3/envs/biopy/lib/python3.11/site-packages/Bio/GenBank/Scanner.py:1528: BiopythonParserWarning: Attempting to parse malformed locus line:
'LOCUS       pUC19-J23100        2686 bp DNA     circular SYN 01-JUN-2024\n'
Found locus 'pUC19-J23100' size '2686' residue_type 'DNA'
Some fields may be wrong.
  warnings.warn(
/home/haseong/anaconda3/envs/biopy/lib/python3.11/site-packages/Bio/GenBank/Scanner.py:1528: BiopythonParserWarning: Attempting to parse malformed locus line:
'LOCUS       pUC19-L2U3H03        2688 bp DNA     circular SYN 01-JUN-2024\n'
Found locus 'pUC19-L2U3H03' size '2688' residue_type 'DNA'
Some fields may be wrong.
  warnings.warn(
/home/haseong/anaconda3/envs/biopy/lib/python3.11/site-packages/Bio/GenBank/Scanner.py:1528: BiopythonParserWarning: Attempting to parse malformed locus line:
'LOCUS       pUC19-RB0030        2750 bp DNA     circular SYN 01-JUN-2024\n'
Found locus 'pUC19-RB0030' size '2750' residue_type 'DNA'
Some fields may be wrong.
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Ft#</th>
<th data-quarto-table-cell-role="th">Label or Note</th>
<th data-quarto-table-cell-role="th">Dir</th>
<th data-quarto-table-cell-role="th">Sta</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Len</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">orf?</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>nd</td>
<td>--&gt;</td>
<td>0</td>
<td>2686</td>
<td>2686</td>
<td>source</td>
<td>no</td>
</tr>
<tr class="even">
<td>1</td>
<td>L:pBR322ori-F</td>
<td>--&gt;</td>
<td>117</td>
<td>137</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>2</td>
<td>L:L4440</td>
<td>--&gt;</td>
<td>370</td>
<td>388</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>3</td>
<td>L:CAP binding si</td>
<td>--&gt;</td>
<td>504</td>
<td>526</td>
<td>22</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>4</td>
<td>L:lac promoter</td>
<td>--&gt;</td>
<td>540</td>
<td>571</td>
<td>31</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="even">
<td>5</td>
<td>L:lac operator</td>
<td>--&gt;</td>
<td>578</td>
<td>595</td>
<td>17</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>6</td>
<td>L:M13/pUC Revers</td>
<td>--&gt;</td>
<td>583</td>
<td>606</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>7</td>
<td>L:M13 rev</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>8</td>
<td>L:M13 Reverse</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>9</td>
<td>L:lacZ-alpha</td>
<td>--&gt;</td>
<td>614</td>
<td>938</td>
<td>324</td>
<td>CDS</td>
<td>no</td>
</tr>
<tr class="odd">
<td>10</td>
<td>L:BsaI-F</td>
<td>--&gt;</td>
<td>631</td>
<td>638</td>
<td>7</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>11</td>
<td>L:oh-v-f1-F</td>
<td>--&gt;</td>
<td>638</td>
<td>642</td>
<td>4</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="odd">
<td>12</td>
<td>L:﻿BBa_J23100</td>
<td>--&gt;</td>
<td>642</td>
<td>677</td>
<td>35</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="even">
<td>13</td>
<td>L:oh-f1-f2-F</td>
<td>--&gt;</td>
<td>677</td>
<td>681</td>
<td>4</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="odd">
<td>14</td>
<td>L:BsaI-R</td>
<td>--&gt;</td>
<td>681</td>
<td>688</td>
<td>7</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>15</td>
<td>L:M13 Forward</td>
<td>&lt;--</td>
<td>688</td>
<td>706</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>16</td>
<td>L:M13 fwd</td>
<td>&lt;--</td>
<td>688</td>
<td>705</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>17</td>
<td>L:M13/pUC Forwar</td>
<td>&lt;--</td>
<td>697</td>
<td>720</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>18</td>
<td>L:pRS-marker</td>
<td>&lt;--</td>
<td>913</td>
<td>933</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>19</td>
<td>L:pGEX 3'</td>
<td>--&gt;</td>
<td>1032</td>
<td>1055</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>20</td>
<td>L:pBRforEco</td>
<td>&lt;--</td>
<td>1092</td>
<td>1111</td>
<td>19</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>21</td>
<td>L:AmpR promoter</td>
<td>--&gt;</td>
<td>1178</td>
<td>1283</td>
<td>105</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="odd">
<td>22</td>
<td>L:AmpR</td>
<td>--&gt;</td>
<td>1283</td>
<td>2144</td>
<td>861</td>
<td>CDS</td>
<td>yes</td>
</tr>
<tr class="even">
<td>23</td>
<td>L:Amp-R</td>
<td>&lt;--</td>
<td>1501</td>
<td>1521</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>24</td>
<td>L:ori</td>
<td>--&gt;</td>
<td>0</td>
<td>2686</td>
<td>589</td>
<td>rep_origin</td>
<td>no</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Ft#</th>
<th data-quarto-table-cell-role="th">Label or Note</th>
<th data-quarto-table-cell-role="th">Dir</th>
<th data-quarto-table-cell-role="th">Sta</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Len</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">orf?</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>nd</td>
<td>--&gt;</td>
<td>0</td>
<td>2750</td>
<td>2750</td>
<td>source</td>
<td>no</td>
</tr>
<tr class="even">
<td>1</td>
<td>L:pBR322ori-F</td>
<td>--&gt;</td>
<td>117</td>
<td>137</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>2</td>
<td>L:L4440</td>
<td>--&gt;</td>
<td>370</td>
<td>388</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>3</td>
<td>L:CAP binding si</td>
<td>--&gt;</td>
<td>504</td>
<td>526</td>
<td>22</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>4</td>
<td>L:lac promoter</td>
<td>--&gt;</td>
<td>540</td>
<td>571</td>
<td>31</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="even">
<td>5</td>
<td>L:lac operator</td>
<td>--&gt;</td>
<td>578</td>
<td>595</td>
<td>17</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>6</td>
<td>L:M13/pUC Revers</td>
<td>--&gt;</td>
<td>583</td>
<td>606</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>7</td>
<td>L:M13 rev</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>8</td>
<td>L:M13 Reverse</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>9</td>
<td>L:BsaI-F</td>
<td>--&gt;</td>
<td>631</td>
<td>638</td>
<td>7</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="odd">
<td>10</td>
<td>L:oh-f1-f2-F</td>
<td>--&gt;</td>
<td>638</td>
<td>642</td>
<td>4</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>11</td>
<td>L:R_BBa_B0030</td>
<td>--&gt;</td>
<td>642</td>
<td>741</td>
<td>99</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="odd">
<td>12</td>
<td>L:sTRSV HHRz</td>
<td>--&gt;</td>
<td>644</td>
<td>694</td>
<td>50</td>
<td>misc_RNA</td>
<td>no</td>
</tr>
<tr class="even">
<td>13</td>
<td>N:strong bacteri</td>
<td>--&gt;</td>
<td>727</td>
<td>739</td>
<td>12</td>
<td>RBS</td>
<td>no</td>
</tr>
<tr class="odd">
<td>14</td>
<td>L:oh-f2-f3-F</td>
<td>--&gt;</td>
<td>741</td>
<td>745</td>
<td>4</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>15</td>
<td>L:BsaI-R</td>
<td>--&gt;</td>
<td>745</td>
<td>752</td>
<td>7</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="odd">
<td>16</td>
<td>L:M13 Forward</td>
<td>&lt;--</td>
<td>752</td>
<td>770</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>17</td>
<td>L:M13 fwd</td>
<td>&lt;--</td>
<td>752</td>
<td>769</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>18</td>
<td>L:M13/pUC Forwar</td>
<td>&lt;--</td>
<td>761</td>
<td>784</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>19</td>
<td>L:pRS-marker</td>
<td>&lt;--</td>
<td>977</td>
<td>997</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>20</td>
<td>L:pGEX 3'</td>
<td>--&gt;</td>
<td>1096</td>
<td>1119</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>21</td>
<td>L:pBRforEco</td>
<td>&lt;--</td>
<td>1156</td>
<td>1175</td>
<td>19</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>22</td>
<td>L:AmpR promoter</td>
<td>--&gt;</td>
<td>1242</td>
<td>1347</td>
<td>105</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="even">
<td>23</td>
<td>L:AmpR</td>
<td>--&gt;</td>
<td>1347</td>
<td>2208</td>
<td>861</td>
<td>CDS</td>
<td>yes</td>
</tr>
<tr class="odd">
<td>24</td>
<td>L:Amp-R</td>
<td>&lt;--</td>
<td>1565</td>
<td>1585</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>25</td>
<td>L:ori</td>
<td>--&gt;</td>
<td>0</td>
<td>2750</td>
<td>589</td>
<td>rep_origin</td>
<td>no</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Ft#</th>
<th data-quarto-table-cell-role="th">Label or Note</th>
<th data-quarto-table-cell-role="th">Dir</th>
<th data-quarto-table-cell-role="th">Sta</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Len</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">orf?</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>nd</td>
<td>--&gt;</td>
<td>0</td>
<td>3371</td>
<td>3371</td>
<td>source</td>
<td>no</td>
</tr>
<tr class="even">
<td>1</td>
<td>L:pBR322ori-F</td>
<td>--&gt;</td>
<td>117</td>
<td>137</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>2</td>
<td>L:L4440</td>
<td>--&gt;</td>
<td>370</td>
<td>388</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>3</td>
<td>L:CAP binding si</td>
<td>--&gt;</td>
<td>504</td>
<td>526</td>
<td>22</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>4</td>
<td>L:lac promoter</td>
<td>--&gt;</td>
<td>540</td>
<td>571</td>
<td>31</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="even">
<td>5</td>
<td>L:lac operator</td>
<td>--&gt;</td>
<td>578</td>
<td>595</td>
<td>17</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>6</td>
<td>L:M13/pUC Revers</td>
<td>--&gt;</td>
<td>583</td>
<td>606</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>7</td>
<td>L:M13 rev</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>8</td>
<td>L:M13 Reverse</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>9</td>
<td>L:lacZ-alpha</td>
<td>--&gt;</td>
<td>614</td>
<td>1623</td>
<td>1009</td>
<td>CDS</td>
<td>no</td>
</tr>
<tr class="odd">
<td>10</td>
<td>L:BsaI-F</td>
<td>--&gt;</td>
<td>631</td>
<td>638</td>
<td>7</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>11</td>
<td>L:lacZ-alpha</td>
<td>--&gt;</td>
<td>638</td>
<td>642</td>
<td>4</td>
<td>CDS</td>
<td>no</td>
</tr>
<tr class="odd">
<td>12</td>
<td>L:oh-f2-f3-F</td>
<td>--&gt;</td>
<td>638</td>
<td>642</td>
<td>4</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>13</td>
<td>L:EGFP</td>
<td>--&gt;</td>
<td>642</td>
<td>1362</td>
<td>720</td>
<td>CDS</td>
<td>yes</td>
</tr>
<tr class="odd">
<td>14</td>
<td>L:EGFP</td>
<td>--&gt;</td>
<td>642</td>
<td>1359</td>
<td>717</td>
<td>CDS</td>
<td>no</td>
</tr>
<tr class="even">
<td>15</td>
<td>L:EGFP-N</td>
<td>&lt;--</td>
<td>687</td>
<td>709</td>
<td>22</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>16</td>
<td>L:EXFP-R</td>
<td>&lt;--</td>
<td>948</td>
<td>968</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>17</td>
<td>L:EGFP-C</td>
<td>--&gt;</td>
<td>1295</td>
<td>1317</td>
<td>22</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>18</td>
<td>L:oh-f3-f4-F</td>
<td>--&gt;</td>
<td>1362</td>
<td>1366</td>
<td>4</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>19</td>
<td>L:BsaI-R</td>
<td>--&gt;</td>
<td>1366</td>
<td>1373</td>
<td>7</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="odd">
<td>20</td>
<td>L:M13 Forward</td>
<td>&lt;--</td>
<td>1373</td>
<td>1391</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>21</td>
<td>L:M13 fwd</td>
<td>&lt;--</td>
<td>1373</td>
<td>1390</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>22</td>
<td>L:M13/pUC Forwar</td>
<td>&lt;--</td>
<td>1382</td>
<td>1405</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>23</td>
<td>L:pRS-marker</td>
<td>&lt;--</td>
<td>1598</td>
<td>1618</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>24</td>
<td>L:pGEX 3'</td>
<td>--&gt;</td>
<td>1717</td>
<td>1740</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>25</td>
<td>L:pBRforEco</td>
<td>&lt;--</td>
<td>1777</td>
<td>1796</td>
<td>19</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>26</td>
<td>L:AmpR promoter</td>
<td>--&gt;</td>
<td>1863</td>
<td>1968</td>
<td>105</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="even">
<td>27</td>
<td>L:AmpR</td>
<td>--&gt;</td>
<td>1968</td>
<td>2829</td>
<td>861</td>
<td>CDS</td>
<td>yes</td>
</tr>
<tr class="odd">
<td>28</td>
<td>L:Amp-R</td>
<td>&lt;--</td>
<td>2186</td>
<td>2206</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>29</td>
<td>L:ori</td>
<td>--&gt;</td>
<td>0</td>
<td>3371</td>
<td>589</td>
<td>rep_origin</td>
<td>no</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">Ft#</th>
<th data-quarto-table-cell-role="th">Label or Note</th>
<th data-quarto-table-cell-role="th">Dir</th>
<th data-quarto-table-cell-role="th">Sta</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Len</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">orf?</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>nd</td>
<td>--&gt;</td>
<td>0</td>
<td>2688</td>
<td>2688</td>
<td>source</td>
<td>no</td>
</tr>
<tr class="even">
<td>1</td>
<td>L:pBR322ori-F</td>
<td>--&gt;</td>
<td>117</td>
<td>137</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>2</td>
<td>L:L4440</td>
<td>--&gt;</td>
<td>370</td>
<td>388</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>3</td>
<td>L:CAP binding si</td>
<td>--&gt;</td>
<td>504</td>
<td>526</td>
<td>22</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>4</td>
<td>L:lac promoter</td>
<td>--&gt;</td>
<td>540</td>
<td>571</td>
<td>31</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="even">
<td>5</td>
<td>L:lac operator</td>
<td>--&gt;</td>
<td>578</td>
<td>595</td>
<td>17</td>
<td>protein_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>6</td>
<td>L:M13/pUC Revers</td>
<td>--&gt;</td>
<td>583</td>
<td>606</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>7</td>
<td>L:M13 rev</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>8</td>
<td>L:M13 Reverse</td>
<td>--&gt;</td>
<td>602</td>
<td>619</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>9</td>
<td>L:BsaI-F</td>
<td>--&gt;</td>
<td>631</td>
<td>638</td>
<td>7</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="odd">
<td>10</td>
<td>L:oh-f3-f4-F</td>
<td>--&gt;</td>
<td>638</td>
<td>642</td>
<td>4</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>11</td>
<td>L:L2U3H03</td>
<td>--&gt;</td>
<td>642</td>
<td>679</td>
<td>37</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="odd">
<td>12</td>
<td>L:oh-f7-v-F</td>
<td>--&gt;</td>
<td>679</td>
<td>683</td>
<td>4</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="even">
<td>13</td>
<td>L:BsaI-R</td>
<td>--&gt;</td>
<td>683</td>
<td>690</td>
<td>7</td>
<td>misc_feature</td>
<td>no</td>
</tr>
<tr class="odd">
<td>14</td>
<td>L:M13 Forward</td>
<td>&lt;--</td>
<td>690</td>
<td>708</td>
<td>18</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>15</td>
<td>L:M13 fwd</td>
<td>&lt;--</td>
<td>690</td>
<td>707</td>
<td>17</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>16</td>
<td>L:M13/pUC Forwar</td>
<td>&lt;--</td>
<td>699</td>
<td>722</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>17</td>
<td>L:pRS-marker</td>
<td>&lt;--</td>
<td>915</td>
<td>935</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>18</td>
<td>L:pGEX 3'</td>
<td>--&gt;</td>
<td>1034</td>
<td>1057</td>
<td>23</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>19</td>
<td>L:pBRforEco</td>
<td>&lt;--</td>
<td>1094</td>
<td>1113</td>
<td>19</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="odd">
<td>20</td>
<td>L:AmpR promoter</td>
<td>--&gt;</td>
<td>1180</td>
<td>1285</td>
<td>105</td>
<td>promoter</td>
<td>no</td>
</tr>
<tr class="even">
<td>21</td>
<td>L:AmpR</td>
<td>--&gt;</td>
<td>1285</td>
<td>2146</td>
<td>861</td>
<td>CDS</td>
<td>yes</td>
</tr>
<tr class="odd">
<td>22</td>
<td>L:Amp-R</td>
<td>&lt;--</td>
<td>1503</td>
<td>1523</td>
<td>20</td>
<td>primer_bind</td>
<td>no</td>
</tr>
<tr class="even">
<td>23</td>
<td>L:ori</td>
<td>--&gt;</td>
<td>0</td>
<td>2688</td>
<td>589</td>
<td>rep_origin</td>
<td>no</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<ul>
<li>Make a function for searching by label and returning positions</li>
</ul>
<div id="cell-39" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co">## search by label and return the location</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_positions_by_label(record, label) :</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    feature_list <span class="op">=</span> record.features</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    pos <span class="op">=</span> {<span class="st">'start'</span>: <span class="va">None</span>, <span class="st">'end'</span>: <span class="va">None</span>}</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> feature <span class="kw">in</span> feature_list:</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="st">"label"</span> <span class="kw">in</span> feature.qualifiers:</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> feature.qualifiers[<span class="st">'label'</span>] <span class="op">==</span> [label]:</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>                pos[<span class="st">'start'</span>] <span class="op">=</span> <span class="bu">int</span>(feature.location.start)</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>                pos[<span class="st">'end'</span>] <span class="op">=</span> <span class="bu">int</span>(feature.location.end)</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> pos</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_record_between_labels(record, label1, label2) :</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>    pos <span class="op">=</span> {<span class="st">'start'</span>: get_positions_by_label(record, label1)[<span class="st">'start'</span>], <span class="st">'end'</span>: get_positions_by_label(record, label2)[<span class="st">'end'</span>]}</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> pos[<span class="st">'start'</span>] <span class="op">==</span> <span class="va">None</span> <span class="kw">or</span> pos[<span class="st">'end'</span>] <span class="op">==</span> <span class="va">None</span>:</span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"label not found"</span>)</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">None</span></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> record[pos[<span class="st">'start'</span>]:pos[<span class="st">'end'</span>]]</span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>target_record <span class="op">=</span> get_record_between_labels(egfp, <span class="st">"BsaI-F"</span>, <span class="st">"BsaI-R"</span>)</span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a><span class="co"># print("pos:", target_record)</span></span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>display(target_record.figure())</span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>target_record.seq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div class="ansi-escaped-output">
<pre>Dseqrecord(-742)
<span class="ansi-bright-yellow-bg">GGTCTCA</span>GTCAATGGTGAGCAAGGGCGAGGAGCTGTTCACCGGGGTGGTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCGTGCCCTGGCCCACCCTCGTGACCACCCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGCCCGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGCAACTACAAGACCCGCGCCGAGGTGAAGTTCGAGGGCGACACCCTGGTGAACCGCATCGAGCTGAAGGGCATCGACTTCAAGGAGGACGGCAACATCCTGGGGCACAAGCTGGAGTACAACTACAACAGCCACAACGTCTATATCATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCGACGGCCCCGTGCTGCTGCCCGACAACCACTACCTGAGCACCCAGTCCGCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACAAGTAAGGGATGAGACC
CCAGAGTCAGTTACCACTCGTTCCCGCTCCTCGACAAGTGGCCCCACCACGGGTAGGACCAGCTCGACCTGCCGCTGCATTTGCCGGTGTTCAAGTCGCACAGGCCGCTCCCGCTCCCGCTACGGTGGATGCCGTTCGACTGGGACTTCAAGTAGACGTGGTGGCCGTTCGACGGGCACGGGACCGGGTGGGAGCACTGGTGGGACTGGATGCCGCACGTCACGAAGTCGGCGATGGGGCTGGTGTACTTCGTCGTGCTGAAGAAGTTCAGGCGGTACGGGCTTCCGATGCAGGTCCTCGCGTGGTAGAAGAAGTTCCTGCTGCCGTTGATGTTCTGGGCGCGGCTCCACTTCAAGCTCCCGCTGTGGGACCACTTGGCGTAGCTCGACTTCCCGTAGCTGAAGTTCCTCCTGCCGTTGTAGGACCCCGTGTTCGACCTCATGTTGATGTTGTCGGTGTTGCAGATATAGTACCGGCTGTTCGTCTTCTTGCCGTAGTTCCACTTGAAGTTCTAGGCGGTGTTGTAGCTCCTGCCGTCGCACGTCGAGCGGCTGGTGATGGTCGTCTTGTGGGGGTAGCCGCTGCCGGGGCACGACGACGGGCTGTTGGTGATGGACTCGTGGGTCAGGCGGGACTCGTTTCTGGGGTTGCTCTTCGCGCTAGTGTACCAGGACGACCTCAAGCACTGGCGGCGGCCCTAGTGAGAGCCGTACCTGCTCGACATGTTCATTCCCTACTCTGG</pre>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="35">
<pre><code>Dseq(-742)
GGTC..GACC
CCAG..CTGG</code></pre>
</div>
</div>
</section>
<section id="part-preparation-by-pcr" class="level3" data-number="2.8.2">
<h3 data-number="2.8.2" class="anchored" data-anchor-id="part-preparation-by-pcr"><span class="header-section-number">2.8.2</span> Part preparation by PCR</h3>
<ul>
<li>from “BsaI-F” to “BsaI-R”</li>
</ul>
<div id="cell-41" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.readers <span class="im">import</span> read</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> primers <span class="im">import</span> create, primers</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.<span class="bu">all</span> <span class="im">import</span> pcr</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>label1 <span class="op">=</span> <span class="st">"BsaI-F"</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>label2 <span class="op">=</span> <span class="st">"BsaI-R"</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>promoter_target <span class="op">=</span> get_record_between_labels(promoter, label1, label2)</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>rbs_target <span class="op">=</span> get_record_between_labels(rbs, label1, label2)</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>egfp_target <span class="op">=</span> get_record_between_labels(egfp, label1, label2)</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>terminator_target <span class="op">=</span> get_record_between_labels(terminator, label1, label2)</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>fwd, rev <span class="op">=</span> create(<span class="bu">str</span>(promoter_target.seq))</span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>promoter_pcr_product <span class="op">=</span> pcr(fwd.seq, rev.seq, promoter)</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>display(promoter_pcr_product.figure())</span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a>fwd, rev <span class="op">=</span> create(<span class="bu">str</span>(rbs_target.seq))</span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a>rbs_pcr_product <span class="op">=</span> pcr(fwd.seq, rev.seq, rbs)</span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a>display(rbs_pcr_product.figure())</span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a>fwd, rev <span class="op">=</span> create(<span class="bu">str</span>(egfp_target.seq))</span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a>egfp_pcr_product <span class="op">=</span> pcr(fwd.seq, rev.seq, egfp)</span>
<span id="cb36-23"><a href="#cb36-23" aria-hidden="true" tabindex="-1"></a>display(egfp_pcr_product.figure())</span>
<span id="cb36-24"><a href="#cb36-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-25"><a href="#cb36-25" aria-hidden="true" tabindex="-1"></a>fwd, rev <span class="op">=</span> create(<span class="bu">str</span>(terminator_target.seq))</span>
<span id="cb36-26"><a href="#cb36-26" aria-hidden="true" tabindex="-1"></a>terminator_pcr_product <span class="op">=</span> pcr(fwd.seq, rev.seq, terminator)</span>
<span id="cb36-27"><a href="#cb36-27" aria-hidden="true" tabindex="-1"></a>display(terminator_pcr_product.figure())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>5GGTCTCAGTCAATGGTGA...TACAAGTAAGGGATGAGACC3
                      ||||||||||||||||||||
                     3ATGTTCATTCCCTACTCTGG5
5GGTCTCAGTCAATGGTGA3
 ||||||||||||||||||
3CCAGAGTCAGTTACCACT...ATGTTCATTCCCTACTCTGG5</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>5GGTCTCAAAGCTTGACG...CTAGCCTCCAGAGACC3
                     ||||||||||||||||
                    3GATCGGAGGTCTCTGG5
5GGTCTCAAAGCTTGACG3
 |||||||||||||||||
3CCAGAGTTTCGAACTGC...GATCGGAGGTCTCTGG5</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>5GGTCTCACTCCAGCTG...GAGGAGAAATAGTCATGAGACC3
                    ||||||||||||||||||||||
                   3CTCCTCTTTATCAGTACTCTGG5
5GGTCTCACTCCAGCTG3
 ||||||||||||||||
3CCAGAGTGAGGTCGAC...CTCCTCTTTATCAGTACTCTGG5</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>5GGTCTCAGGGATAGCG...TTGTTGAGCGAATGAGACC3
                    |||||||||||||||||||
                   3AACAACTCGCTTACTCTGG5
5GGTCTCAGGGATAGCG3
 ||||||||||||||||
3CCAGAGTCCCTATCGC...AACAACTCGCTTACTCTGG5</code></pre>
</div>
</div>
</section>
<section id="enzyme-cut-and-assembly" class="level3" data-number="2.8.3">
<h3 data-number="2.8.3" class="anchored" data-anchor-id="enzyme-cut-and-assembly"><span class="header-section-number">2.8.3</span> Enzyme cut and assembly</h3>
<div id="cell-43" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> Bio.Restriction <span class="im">import</span> BsaI</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.assembly <span class="im">import</span> Assembly</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pydna.common_sub_strings <span class="im">import</span> terminal_overlap</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>promoter_cut_product <span class="op">=</span> rbs_pcr_product.cut(BsaI)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(promoter_cut_product))</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>display(promoter_cut_product[<span class="dv">1</span>].figure())</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>rbs_cut_product <span class="op">=</span> rbs_pcr_product.cut(BsaI)</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(rbs_cut_product))</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>display(rbs_cut_product[<span class="dv">1</span>].figure())</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>egfp_cut_product <span class="op">=</span> egfp_pcr_product.cut(BsaI)</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(egfp_cut_product))</span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>display(egfp_cut_product[<span class="dv">1</span>].figure())</span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>terminator_cut_product <span class="op">=</span> terminator_pcr_product.cut(BsaI)</span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(terminator_cut_product))</span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>display(terminator_cut_product[<span class="dv">1</span>].figure())</span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a>asm <span class="op">=</span> Assembly((promoter_cut_product[<span class="dv">1</span>], rbs_cut_product[<span class="dv">1</span>], egfp_cut_product[<span class="dv">1</span>], terminator_cut_product[<span class="dv">1</span>]), algorithm <span class="op">=</span> terminal_overlap, limit <span class="op">=</span> <span class="dv">4</span>)</span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a>candidate <span class="op">=</span> asm.assemble_linear()</span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(candidate))</span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-27"><a href="#cb41-27" aria-hidden="true" tabindex="-1"></a>candidate[<span class="dv">0</span>].figure()   </span>
<span id="cb41-28"><a href="#cb41-28" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>3
3
3
3
1</code></pre>
</div>
<div class="cell-output cell-output-display">
<div class="ansi-escaped-output">
<pre>Dseqrecord(-107)
<span class="ansi-bright-yellow-bg">CTCC</span>AGCTGTCACCGGATGTGCTTTCCGGTCTGATGAGTCCGTGAGGACGAAACAGCCTCTACAAATAATTTTGTTTAATCTAGAGATTAAAGAGGAGAAATA    
    TCGACAGTGGCCTACACGAAAGGCCAGACTACTCAGGCACTCCTGCTTTGTCGGAGATGTTTATTAAAACAAATTAGATCTCTAATTTCTCCTCTTTATCAGT</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div class="ansi-escaped-output">
<pre>Dseqrecord(-107)
<span class="ansi-bright-yellow-bg">CTCC</span>AGCTGTCACCGGATGTGCTTTCCGGTCTGATGAGTCCGTGAGGACGAAACAGCCTCTACAAATAATTTTGTTTAATCTAGAGATTAAAGAGGAGAAATA    
    TCGACAGTGGCCTACACGAAAGGCCAGACTACTCAGGCACTCCTGCTTTGTCGGAGATGTTTATTAAAACAAATTAGATCTCTAATTTCTCCTCTTTATCAGT</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div class="ansi-escaped-output">
<pre>Dseqrecord(-728)
<span class="ansi-bright-yellow-bg">GTCA</span>ATGGTGAGCAAGGGCGAGGAGCTGTTCACCGGGGTGGTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCGTGCCCTGGCCCACCCTCGTGACCACCCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGCCCGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGCAACTACAAGACCCGCGCCGAGGTGAAGTTCGAGGGCGACACCCTGGTGAACCGCATCGAGCTGAAGGGCATCGACTTCAAGGAGGACGGCAACATCCTGGGGCACAAGCTGGAGTACAACTACAACAGCCACAACGTCTATATCATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCGACGGCCCCGTGCTGCTGCCCGACAACCACTACCTGAGCACCCAGTCCGCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACAAGTAA    
    TACCACTCGTTCCCGCTCCTCGACAAGTGGCCCCACCACGGGTAGGACCAGCTCGACCTGCCGCTGCATTTGCCGGTGTTCAAGTCGCACAGGCCGCTCCCGCTCCCGCTACGGTGGATGCCGTTCGACTGGGACTTCAAGTAGACGTGGTGGCCGTTCGACGGGCACGGGACCGGGTGGGAGCACTGGTGGGACTGGATGCCGCACGTCACGAAGTCGGCGATGGGGCTGGTGTACTTCGTCGTGCTGAAGAAGTTCAGGCGGTACGGGCTTCCGATGCAGGTCCTCGCGTGGTAGAAGAAGTTCCTGCTGCCGTTGATGTTCTGGGCGCGGCTCCACTTCAAGCTCCCGCTGTGGGACCACTTGGCGTAGCTCGACTTCCCGTAGCTGAAGTTCCTCCTGCCGTTGTAGGACCCCGTGTTCGACCTCATGTTGATGTTGTCGGTGTTGCAGATATAGTACCGGCTGTTCGTCTTCTTGCCGTAGTTCCACTTGAAGTTCTAGGCGGTGTTGTAGCTCCTGCCGTCGCACGTCGAGCGGCTGGTGATGGTCGTCTTGTGGGGGTAGCCGCTGCCGGGGCACGACGACGGGCTGTTGGTGATGGACTCGTGGGTCAGGCGGGACTCGTTTCTGGGGTTGCTCTTCGCGCTAGTGTACCAGGACGACCTCAAGCACTGGCGGCGGCCCTAGTGAGAGCCGTACCTGCTCGACATGTTCATTCCCT</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div class="ansi-escaped-output">
<pre>Dseqrecord(-45)
<span class="ansi-bright-yellow-bg">GGGA</span>TAGCGTGACCGGCGCATCGGTCACGCTATTTGTTGAG    
    ATCGCACTGGCCGCGTAGCCAGTGCGATAAACAACTCGCTT</pre>
</div>
</div>
<div class="cell-output cell-output-display" data-execution_count="37">
<pre><code>name| 4
     \/
     /\
      4|name| 4
             \/
             /\
              4|name</code></pre>
</div>
</div>
<div id="cell-44" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>promoter_cut_product[<span class="dv">1</span>].figure()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<div class="ansi-escaped-output">
<pre>Dseqrecord(-107)
<span class="ansi-bright-yellow-bg">CTCC</span>AGCTGTCACCGGATGTGCTTTCCGGTCTGATGAGTCCGTGAGGACGAAACAGCCTCTACAAATAATTTTGTTTAATCTAGAGATTAAAGAGGAGAAATA    
    TCGACAGTGGCCTACACGAAAGGCCAGACTACTCAGGCACTCCTGCTTTGTCGGAGATGTTTATTAAAACAAATTAGATCTCTAATTTCTCCTCTTTATCAGT</pre>
</div>
</div>
</div>
</section>
</section>
<section id="assembly-with-wrong-fragments" class="level2" data-number="2.9">
<h2 data-number="2.9" class="anchored" data-anchor-id="assembly-with-wrong-fragments"><span class="header-section-number">2.9</span> Assembly with wrong fragments</h2>
<div id="cell-46" class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>promoter <span class="op">=</span> read(<span class="st">"data/parts/pUC19-J23100.gb"</span>)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>terminator <span class="op">=</span> read(<span class="st">"data/parts/pUC19-L2U3H03.gb"</span>)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>rbs <span class="op">=</span> read(<span class="st">"data/parts/pUC19-RB0030.gb"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./index.html" class="pagination-link" aria-label="Home">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Home</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>2024 lecture note written by Haseong Kim</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This book was built with Quarto</p>
</div>
  </div>
</footer>




</body></html>